include/zephyr/device.h

```diff
@@ -750,12 +750,60 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 	return z_device_is_ready(dev);
 }
 
+#ifdef CONFIG_DEVICE_DEFER_INIT
+/**
+ * @brief Initialize a device.
+ *
+ * A device whose initialization was deferred needs to be initialized manually
+ * via this call.
+ *
+ * @param dev device to be initialized.
+ *
+ * @retval -ENOENT If device was not found
+ * @retval -errno For other errors.
+ */
+__syscall int device_init(const struct device *dev);
+
+/**
+ * @brief Possible results for callback called before device initialization.
+ */
+enum device_before_init_result {
+	DEVICE_INIT,
+	DEVICE_DEFER_INIT
+};
```

Comment:
Should we drop this and just return a boolean? maybe rename the field to something like `device_allow_init` or `device_block_init` and adjust accordingly. This is nice and unambiguous but these type of enum return are a bit uncommon in Zephyr and it somehow reminds me of those horrible `return HAL_SOMETHING_OK;` that few private SDK have. :-)

@gmarull or anyone else, any opinion? @cfriedt also suggested that in https://github.com/zephyrproject-rtos/zephyr/pull/67335#issuecomment-1887115782

Reply:
Hmm... yeah, this looks and smells like a `bool`